rir/src/compiler/opt/elide_env_spec.cpp

```diff
@@ -14,61 +14,97 @@ void ElideEnvSpec::apply(RirCompiler&, ClosureVersion* function,
+            fwdCheckpoints.emplace(bb, cp);
+        }
     });
 
     Visitor::run(function->entry, [&](BB* bb) {
-        if (!checkpoints.count(bb))
-            return;
-
-        auto cp = checkpoints.at(bb);
-
+        std::unordered_map<Value*, Value*> stubFor;
         auto ip = bb->begin();
         while (ip != bb->end()) {
             Instruction* i = *ip;
             auto next = ip + 1;
 
-            if (i->envOnlyForObj() && i->hasEnv()) {
-                bool objectExpected = false;
-
-                i->eachArg([&](Value* arg) {
-                    if (arg != i->env())
-                        if (arg->type.maybeObj() &&
-                            arg->typeFeedback.maybeObj())
-                            objectExpected = true;
-                });
+            if (i->hasEnv()) {
+                // Speculatively elide environments on instructions that only
+                // require them in case any argument is an object
+                if (i->envOnlyForObj() && bckCheckpoints.count(bb) &&
+                    nonObjectArgs(i)) {
+                    auto cp = bckCheckpoints.at(bb);
 
-                if (!objectExpected) {
                     i->elideEnv();
                     i->eachArg([&](Value* arg) {
                         if (arg != i->env())
-                            if (arg->type.maybeObj())
-                                ip = insertExpectNotObj(bb, arg, cp, ip);
+                            if (arg->type.maybeObj()) {
+                                auto condition = (new TypeTest(arg))->object();
+                                BBTransform::insertAssume(bb, condition, cp, ip,
+                                                          false);
+                            }
                     });
                     next = ip + 1;
                     i->type.setNotObject();
                 }
+
+                // Speculatively elide environments on force instructions that
+                // only require them in case the force access promises
+                // reflectively
+                if (Force::Cast(*ip) && fwdCheckpoints.count(bb)) {
+                    auto force = Force::Cast(*ip);
+                    auto environment = MkEnv::Cast(force->env());
+                    auto cp = fwdCheckpoints.at(bb);
+                    static std::unordered_set<Tag> forces{Tag::Force,
+                                                          Tag::FrameState};
+
+                    if (!environment->stub &&
+                        environment->usesAreOnly(function->entry, forces)) {
+                        environment = MkEnv::Cast(force->env());
+
+                        if (!stubFor.count(environment)) {
+                            auto stub = (MkEnv*)environment->clone();
+                            stub->stub = true;
+                            stubFor.emplace(environment, stub);
+                            environment->bb()->insert(
```

Comment:
if `environment()->bb() == bb` then this will potentially trash the iterator `it` (because you are inserting into the container you are iterating over).